### PR TITLE
LPS-197482 Grouping Multiple Edges between Objects in Dropdowns

### PR DESCRIPTION
**Pull request by igor-franca**
Related [Ticket](https://liferay.atlassian.net/browse/LPS-197482)

Motivation:

When dealing with lots of connections between two objects, we ended up with too many edges. Keeping them organized became a challenge as their number grew.

To make things easier, we've tweaked the way edges are handled. Now, the system can handle a bunch of connections all at once, making it simpler to manage and maintain consistency.

screenshot-1705686220

### CODE REVIEW

**Comment by rebsilva** [comment]
just started reviewing! :cherry_blossom:

File: modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ModelBuilder/Edges/ObjectRelationshipEdge.tsx [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+				};
+			});
+		}
+	}, [activePopover, selected]);
+
+	return hasManyObjectRelationships ? (
+		<>
+			<path
+				className="react-flow__edge-path"
+				d={edgePath}
```
[the remark this answers is not in the document]
Comment (igor-franca) [reply modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ModelBuilder/Edges/ObjectRelationshipEdge.tsx:109]: Hey @rebsilva, 

> _The d attribute defines a path to be drawn._
> 
> _A path definition is a list of [path commands](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/d#path_commands) where each command is composed of a command letter and numbers that represent the command parameters. The commands are [detailed below](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/d#path_commands)._

Excerpt taken from: https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/d

File: modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ModelBuilder/Edges/ObjectRelationshipEdge.tsx [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+				id={edgeId}
+				style={objectRelationshipEdgeStyle}
+			/>
+
+			<EdgeText
+				label={data.length}
+				labelBgBorderRadius={4}
+				labelBgPadding={[8, 5]}
+				labelBgStyle={labelBgStyle}
+				labelShowBg
```
[the remark this answers is not in the document]
Comment (igor-franca) [reply modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ModelBuilder/Edges/ObjectRelationshipEdge.tsx:119]: Hey @rebsilva,

From here I have to pass the properties expected by the library, the EdgeText component is exported from the library and requests the following properties:

```
export interface EdgeTextProps extends HTMLAttributes<SVGElement> {
    x: number;
    y: number;
    label?: string | ReactNode;
    labelStyle?: CSSProperties;
    labelShowBg?: boolean;
    labelBgStyle?: CSSProperties;
    labelBgPadding?: [number, number];
    labelBgBorderRadius?: number;
}
```

So the best I can do is:

```
<EdgeText
	label={data.length}
	labelBgBorderRadius={4}
	labelBgPadding={[8, 5]}
	labelBgStyle={labelBackgroundStyle}
	labelShowBg
	labelStyle={labelStyle}
	onClick={(event) => {
		triggerElementRef.current = event.target as HTMLElement;

		setActivePopover(!activePopover);
	}}
	x={edgeCenterX}
	y={edgeCenterY}
/>
```

File: modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ModelBuilder/Edges/ObjectRelationshipEdge.tsx [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+	const [_, dispatch] = useObjectFolderContext();
+	const [activePopover, setActivePopover] = useState(false);
+	const [
+		objectRelationshipEdgeStyle,
+		setObjectRelationshipEdgeStyle,
+	] = useState({
+		...style,
+		...getInitialObjectRelationshipEdgeStyle(selected),
+	});
+	const [labelBgStyle, setLabelBgStyle] = useState(
```
[the remark this answers is not in the document]
Comment (igor-franca) [reply modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ModelBuilder/Edges/ObjectRelationshipEdge.tsx:71]: I agree @rebsilva, this is the way the library hopes to receive this property. But we can improve on our side!

File: modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ModelBuilder/Edges/ObjectRelationshipEdge.tsx [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+					style={objectRelationshipEdgeStyle}
+				/>
+			)}
+
+			<EdgeText
+				label={label}
+				labelBgBorderRadius={4}
+				labelBgPadding={[8, 5]}
+				labelBgStyle={labelBgStyle}
+				labelShowBg
```
[the remark this answers is not in the document]
Comment (igor-franca) [reply modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ModelBuilder/Edges/ObjectRelationshipEdge.tsx:200]: [Answered](https://github.com/liferay-objects/liferay-portal/pull/3057#discussion_r1462439321)

File: modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ModelBuilder/Edges/objectRelationshipEdgeFactory.ts [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+	const objectDefinition1ObjectDefinition2ObjectRelationships = objectRelationshipMap.getValueByExternalReferenceCodes(
+		objectRelationship.objectDefinitionExternalReferenceCode1,
+		objectRelationship.objectDefinitionExternalReferenceCode2
+	);
+
+	const objectDefinition2ObjectDefinition1ObjectRelationships = objectRelationshipMap.getValueByExternalReferenceCodes(
+		objectRelationship.objectDefinitionExternalReferenceCode2,
+		objectRelationship.objectDefinitionExternalReferenceCode1
+	);
```
[the remark this answers is not in the document]
Comment (igor-franca) [reply modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ModelBuilder/Edges/objectRelationshipEdgeFactory.ts]: Carol's surveys were great and were crucial to improving the code!

**Comment by igor-franca** [comment]
ci:test:sf

File: modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ModelBuilder/Edges/ObjectRelationshipEdge.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (rebsilva) [reply modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ModelBuilder/Edges/ObjectRelationshipEdge.tsx:109]: awesome! thanks for the info! :heart:

File: modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ModelBuilder/Edges/ObjectRelationshipEdge.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (rebsilva) [reply modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ModelBuilder/Edges/ObjectRelationshipEdge.tsx:119]: i understand now, the property is named like this by the library...so there's almost nothing we can do about this, thanks for explaining!

**Comment by igor-franca** [comment]
ci:test:sf

**Comment by rebsilva** [comment]
testing the PR i've found some unexpected behaviors so i already told @igor-franca to fix them

**Comment by igor-franca** [comment]
> testing the PR i've found some unexpected behaviors so i already told @igor-franca to fix them

Thank you very much for noticing this behavior! Congratulations on the analyzes you have been doing!

**Comment by igor-franca** [comment]
ci:test:sf

**Comment by igor-franca** [comment]
ci:test:object

**Comment by igor-franca** [comment]
ci:test:relevant

**Comment by rebsilva** [comment]
ci:test:relevant

**Review by rebsilva — APPROVED** [review]
LGTM! :tada: (i had already reviewed, just forgot to put this message!)

**Comment by danvb112** [comment]
Sent to: https://github.com/liferay-objects/liferay-portal/pull/3080

**Pull request closed without merging**